src/lib.rs

```diff
@@ -0,0 +1,334 @@
+        ),
+    }
+
+    /*
+    Example
+    -------
+    RLE encoded data is ordered like this (for 16-bit, 3 sample):
+      Segment: 1     | 2     | 3     | 4     | 5     | 6
+               R MSB | R LSB | G MSB | G LSB | B MSB | B LSB
+
+    A segment contains only the MSB or LSB parts of all the sample pixels
+
+    To minimise the amount of array manipulation later, and to make things
+    faster we interleave each segment in a manner consistent with a planar
+    configuration of 1 (and maintain big endian byte ordering):
+      All red samples             | All green samples           | All blue
+      Pxl 1   Pxl 2   ... Pxl N   | Pxl 1   Pxl 2   ... Pxl N   | ...
+      MSB LSB MSB LSB ... MSB LSB | MSB LSB MSB LSB ... MSB LSB | ...
+    */
+
+    // Watch for overflow here
+    // (31 * 3) * (px_per_sample u32) -> requires u64
+    // usize::overflowing_mul()?
+    let expected_length = (
+        px_per_sample * u32::from(bytes_per_pixel * samples_per_px)
+    ) as usize;
+
+    // TODO: does this make sense?
+    //if u64::MAX > usize::MAX {
+    //    return Err(
+    //        String::from(
+    //            "The size of the decoded data is larger than available"
+    //        ).into()
+    //    )
+    //}
+    let mut out = vec![0u8; expected_length];
+
+    // TODO: Add error check that enough data is available in enc[start..end]
+    for sample in 0..samples_per_px {
+        for byte_offset in 0..bytes_per_pixel {
+            let idx = (sample * bytes_per_pixel + byte_offset) as usize;
+            let start = offsets[idx] as usize;
+            let end = offsets[idx + 1] as usize;
+            println!("Start..end {} {}", start, end);
+            let data = <&[u8]>::try_from(&enc[start..end]).unwrap();
+            let segment = _decode_segment(data)?;
+            let segment_len = segment.len();
+            println!("Segment length {}", segment_len);
+            // Check decoded segment length is good
+            if segment_len != px_per_sample as usize {
+                //
+                println!("Bad segment length");
+            }
+            // Hmm, how do we shuffle efficiently here?
+            //out[]
+        }
+    }
+
+    Ok(out)
+}
+
+
+#[pyfunction]
+fn decode_segment<'a>(enc: &[u8], py: Python<'a>) -> PyResult<&'a PyBytes> {
+    /* Return a decoded RLE segment as bytes.
+
+    Parameters
+    ----------
+    enc : bytes
+        The encoded segment.
+
+    Returns
+    -------
+    bytes
+        The decoded segment.
+    */
+    match _decode_segment(enc) {
+        Ok(frame) => return Ok(PyBytes::new(py, &frame[..])),
+        Err(err) => return Err(PyValueError::new_err(err.to_string())),
+    }
+}
+
+
+fn _decode_segment(enc: &[u8]) -> Result<Vec<u8>, Box<dyn Error>> {
+    /* Return a decoded RLE segment as bytes.
+
+    Parameters
+    ----------
+    enc
+        The encoded segment.
+    */
+    let mut out: Vec<u8> = Vec::new();
```

Comment:
Think about re-working to pre-allocate with `Vec::with_capacity(expected_segment_length)`

Should be faster?